File: server/fall_completion_test.go
package main

import "testing"

func TestFallCompletionFrames(t *testing.T) {
	for _, fatal := range []bool{false, true} {
		for _, restart := range []bool{false, true} {
			s := initial()
			for i := range s.Players {
				s.Players[i].Stun = 100
			}
			for i := range s.Pickups {
				s.Pickups[i].Wait = 100
			}
			p := &s.Players[16]
			p.X, p.Z, p.Stun = 0, 0, 0
			if fatal {
				p.Health = 1
			}
			if !s.damage(7, 16) {
				t.Fatal("fixture damage")
			}
			if restart {
				s.beginRestart(0)
			}
			for frame := 0; frame < 26; frame++ {
				s.simulate(simulationStep, [2]Input{}, [2]bool{true, true})
				if p.physicalFrame != frame {
					t.Fatal("frame", fatal, restart, frame, p.physicalFrame)
				}
				if frame < 25 && (p.Action != 4 || p.Injury != 0 || p.Stun <= 0) {
					t.Fatal("early completion", fatal, restart, frame)
				}
			}
			if fatal {
				if p.Injury != 1 || s.Score[0] != 10 {
					t.Fatal("fatal completion")
				}
			} else if p.Action != 0 || p.poseCursor != 0 || p.Stun != 0 || p.moveX != 0 || p.moveZ != 0 {
				t.Fatal("healthy completion")
			}
		}
	}
}

File: src/fall-completion.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial, step, simulationStep } from "./game.js";
import { damage } from "./features.js";
import { beginRestart } from "./restart.js";

// Source anim_tackled has 26 displayed entries before its first -5.
for (const fatal of [false, true]) for (const restart of [false, true]) {
  test(`fall displays all 26 frames before completion (fatal=${fatal}, restart=${restart})`, () => {
    const s = initial();
    s.players.forEach(p => { p.stun = 100; });
    s.pickups.forEach(p => { p.wait = 100; });
    const p = s.players[16];
    Object.assign(p, { x: 0, z: 0, stun: 0, health: fatal ? 1 : 100 });
    assert.equal(damage(s, 7, 16), true);
    if (restart) beginRestart(s);
    for (let frame = 0; frame < 26; frame++) {
      step(s, simulationStep, {}, [true, true]);
      assert.equal(p.physicalFrame, frame);
      if (frame < 25) {
        assert.equal(p.action, 4);
        assert.equal(p.injury, 0);
        assert.ok(p.stun > 0);
      }
    }
    assert.equal(p.injury, fatal ? 1 : 0);
    assert.equal(s.score[0], fatal ? 10 : 0);
    if (!fatal) {
      assert.equal(p.action, 0);
      assert.equal(p.stun, 0);
      assert.equal(p.poseCursor, 0);
      assert.equal(p.moveX, 0);
      assert.equal(p.moveZ, 0);
    }
  });
}
